common/beerocks/bwl/shared/netlink_socket.cpp

```diff
@@ -0,0 +1,147 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 MaxLinear
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "netlink_socket.h"
+
+#include <netlink/msg.h>
+#include <netlink/netlink.h>
+
+#include <easylogging++.h>
+
+namespace bwl {
+
+netlink_socket::netlink_socket(int protocol) : m_protocol(protocol)
+{
+    // Allocate the netlink socket structure
+    m_nl_socket = nl_socket_alloc();
+    if (!m_nl_socket) {
+        LOG(ERROR) << "Failed to allocate netlink socket!";
+    }
+}
+
+netlink_socket::~netlink_socket()
+{
+    // Deallocate the netlink socket structure
+    if (m_nl_socket) {
+        nl_socket_free(m_nl_socket);
+    }
+}
+
+bool netlink_socket::connect()
+{
```

Comment:
I'd rather leave it as it is. Although it might never be the case, you could create an instance of `netlink_socket` once and later call `connect` and `close` as many times you wanted. But most important of all is that derived classes can configure the socket in their constructor **before** the call to `connect` can ever be made. That's just the case with derived class `nl80211_socket`, which increases the socket's internal buffer size in its constructor.